apps/src/netsim/NetSimPacketEditor.js

```diff
@@ -0,0 +1,495 @@
+ * @param {function} initialConfig.removePacketCallback
+ * @constructor
+ */
+var NetSimPacketEditor = module.exports = function (initialConfig) {
+
+  /**
+   * @type {jQuery}
+   * @private
+   */
+  this.rootDiv_ = $('<div>').addClass('netsim-packet');
+
+  /** @type {number} */
+  this.toAddress = initialConfig.toAddress || 0;
+  
+  /** @type {number} */
+  this.fromAddress = initialConfig.fromAddress || 0;
+  
+  /** @type {number} */
+  this.packetIndex = initialConfig.packetIndex || 1;
+  
+  /** @type {number} */
+  this.packetCount = initialConfig.packetCount || 1;
+
+  /**
+   * Binary string of message body, live-interpreted to other values.
+   * @type {string}
+   */
+  this.message = initialConfig.message || '';
+
+  /**
+   * Maximum packet length configurable by slider.
+   * @type {Number}
+   * @private
+   */
+  this.maxPacketSize_ = initialConfig.maxPacketSize || Infinity;
+
+  /**
+   * Bits per chunk/byte for parsing and formatting purposes.
+   * @type {number}
+   * @private
+   */
+  this.currentChunkSize_ = initialConfig.chunkSize || BITS_PER_BYTE;
+
+  /**
+   * Which encodings should be visible in the editor.
+   * @type {EncodingType[]}
+   * @private
+   */
+  this.enabledEncodings_ = initialConfig.enabledEncodings || [];
+
+  /**
+   * Method to call in order to remove this packet from its parent.
+   * Function should take this PacketEditor as an argument.
+   * @type {function}
+   * @private
+   */
+  this.removePacketCallback_ = initialConfig.removePacketCallback;
+
+  /**
+   * @type {jQuery}
+   * @private
+   */
+  this.removePacketButton_ = null;
+
+  /**
+   * @type {jQuery}
+   * @private
+   */
+  this.bitCounter_ = null;
+  
+  this.render();
+};
+
+/**
+ * Return root div, for hooking up to a parent element.
+ * @returns {jQuery}
+ */
+NetSimPacketEditor.prototype.getRoot = function () {
+  return this.rootDiv_;
+};
+
+/** Replace contents of our root element with our own markup. */
+NetSimPacketEditor.prototype.render = function () {
+  var newMarkup = $(markup({}));
+  this.rootDiv_.html(newMarkup);
+  this.bindElements_();
+  this.updateFields_();
+  this.removePacketButton_.toggle(this.packetCount > 1);
+  NetSimEncodingControl.hideRowsByEncoding(this.rootDiv_, this.enabledEncodings_);
+};
+
+/**
+ * Focus event handler.  If the target element has a 'watermark' class then
+ * it contains text we intend to clear before any editing occurs.  This
+ * handler clears that text and removes the class.
+ * @param focusEvent
+ */
+var removeWatermark = function (focusEvent) {
+  var target = $(focusEvent.target);
+  if (target.hasClass('watermark')) {
+    target.val('');
+    target.removeClass('watermark');
+  }
+};
+
+/**
+ * Creates a keyPress handler that allows only the given characters to be
+ * typed into a text field.
+ * @param {RegExp} whitelistRegex
+ * @return {function} appropriate to pass to .keypress()
+ */
+var whitelistCharacters = function (whitelistRegex) {
```

Comment:
This is neat!

The name sounds a bit too much like a collection for a generator, maybe something like `keypressHandlerWithWhitelist(whitelistRegex)`?